### PR TITLE
fix: deep evaluate matrix strategy

### PR DESCRIPTION
**Pull request by ChristopherHX**
We need to wait for the previous stage to finish, otherwise we don't know how the final strategy will look like.
Then we need to evaluate all expression nodes of the strategy.matrix property, even the mapping keys.

I also added support for the insert directive and array merge operations, both are undocumented features of the github/actions which are available since 2019.

My tests aren't cover all possibilities, but the fixed issue has a test.

Most likely this change collides with @catthehacker plans of rewriting the yaml parser to actionslint, since I use the current yaml.Node types.

I have no idea how big the regressions regarding parallelism are with this change.

Resolves https://github.com/nektos/act/issues/927

### CODE REVIEW

File: pkg/runner/expression.go [hunk 1]
```diff
@@ -130,6 +132,83 @@ func (ee expressionEvaluator) evaluate(in string, isIfExpression bool) (interfac
...
+			return err
+		}
+		return node.Encode(res)
+	} else if node.Kind == yaml.MappingNode {
+		var m map[interface{}]yaml.Node
+		if err := node.Decode(&m); err != nil {
+			return err
+		}
+		// GitHub has this undocumented feature to merge maps, called insert directive
+		insertDirective := regexp.MustCompile(`\${{\s*insert\s*}}`)
```
Comment (ChristopherHX) [inline pkg/runner/expression.go:159]: Sample usage of `${{insert}}`, key collisions result in template error on github and actions/runner
```yaml
on: push
jobs:
  a:
    strategy:
      matrix:
        a:
        - env:
              key1: val1
              ${{insert}}:
                 key2: val2
              ${{insert}}: ${{fromJSON('{"key3": "val3"}')}}
    runs-on: ubuntu-latest
    steps:
    - run: exit 0
```

File: pkg/runner/expression.go [hunk 2]
```diff
@@ -130,6 +132,97 @@ func (ee expressionEvaluator) evaluate(in string, isIfExpression bool) (interfac
...
+	aout := make([]yaml.Node, 0)
+	for i := range a {
+		v := a[i]
+		// Preserve nested sequences
+		wasseq := v.Kind == yaml.SequenceNode
+		if err := ee.EvaluateYamlNode(&v); err != nil {
+			return err
+		}
+		// GitHub has this undocumented feature to merge sequences / arrays
+		// We have a nested sequence via evaluation, merge the arrays
```
Comment (ChristopherHX) [inline pkg/runner/expression.go:201]: Sample usage of the undocumented array merge operation
```
on: push
jobs:
  a:
    strategy:
      matrix:
        a:
        - env:
              key1: val1
        - ${{fromJSON('[{"key2": "val2"}]')}}
    runs-on: ubuntu-latest
    steps:
    - run: exit 0

File: pkg/runner/expression.go [hunk 3]
```diff
@@ -130,6 +132,97 @@ func (ee expressionEvaluator) evaluate(in string, isIfExpression bool) (interfac
...
+
+func (ee expressionEvaluator) EvaluateYamlNode(node *yaml.Node) error {
+	if node.Kind == yaml.ScalarNode {
+		return ee.evaluateScalarYamlNode(node)
+	} else if node.Kind == yaml.MappingNode {
+		return ee.evaluateMappingYamlNode(node)
+	} else if node.Kind == yaml.SequenceNode {
+		return ee.EvaluateSequenceYamlNode(node)
+	}
+	return nil
```
Comment (catthehacker) [inline pkg/runner/expression.go:223]: `switch` would be better here

Comment (ChristopherHX) [reply pkg/runner/expression.go]: Done

**Comment by ChristopherHX** [comment]
I noticed that my first map approuch of evaluateMappingYamlNode failed with yaml validation errors, if I used `${{insert}}` as yaml key multiple times. It seems to be a somewhat complex change, loosly based on how actions/runner works internally.
More parts of act should migrate to use this method for evaluation e.g.
- job.strategy itself the whole block is evaluatable
- job.env the whole block is evaluatable
- job.defaults the whole block is evaluatable
- job.runs-on
- job.steps.*.with the whole block is evaluatable
- job.steps.*.env the whole block is evaluatable
- job.steps.*.continue-on-error
- job.steps.*.timeout-minutes

**Review by ZauberNerd — APPROVED** [review]
So if I understand it correctly, you take the parsed yaml node, turn it back into its string representation, run `Evaluate` on that and then encode it as a `yaml.Node` again?
Feels a bit hack-y to me, but I think it is fine for now until we have refactored the whole workflow parsing to use actionlint and then we can remove this again.

File: pkg/runner/runner.go [hunk 4]
```diff
@@ -115,59 +115,71 @@ func New(runnerConfig *Config) (Runner, error) {
...
-			}
-
-			b := 0
-			for i, matrix := range matrixes {
-				rc := runner.newRunContext(run, matrix)
-				rc.JobName = rc.Name
-				if len(matrixes) > 1 {
-					rc.Name = fmt.Sprintf("%s-%d", rc.Name, i+1)
+	stagePipeline := make([]common.Executor, 0)
+	for i := range plan.Stages {
```
[the remark this answers is not in the document]
Comment (ChristopherHX) [reply pkg/runner/runner.go:119]: golang reuses the loop variables `s`, `stage`, the linter warns you if you do that after my changes.
So what happens if you reference `s`
- you capture the loop variable in a function
- the loop finishes
- the callbacks with references to the loop variables get called
- `s`, `stage` are always assigned to the last stage.

_I also believed that I can safely reference loop variables in `func(){...}`_

[the remark this answers is not in the document]
Comment (ZauberNerd) [reply pkg/runner/runner.go:119]: Ah, yes. We ran into the same issue here: https://github.com/nektos/act/blob/master/pkg/runner/job_executor_test.go#L147-L155 ;)

**Review by KnisterPeter — APPROVED** [review]
Looks fine from my side

**Review by cplee — APPROVED** [review]
lgtm - thanks

**Comment by ChristopherHX** [comment]
> you take the parsed yaml node, turn it back into its string representation

I only look for expressions in scalar values could be boolean, number or string. Maybe we could read the Value field directly like in if statements. I just find it myself safer to use the decode function.

> run Evaluate on that and then encode it as a yaml.Node again

Everything which isn't an expression is skipped and not encoded again
After evaluate we get a raw boolean, number, string, array or mapping which needs to be encoded as yaml.Node.


For actionslint, we may need to replace all yaml types. I rely much on the current act api with yaml.Node, with a lot of convertion functions to translate job requests of the github actions service into act's layout.

> workflow parsing to use actionlint and then we can remove this again.

I doubt we can remove this later without breaking my test cases.

**Pull request merged**